### ISSUE TITLE
Add tab steps

### ISSUE THREAD

**Issue by denzelem** [makandra/spreewald#126]
What do you think about adding steps to switch between tabs?

```
And I am going to open a new tab
And I close the browser tab
And I open ... in a new tab
And I open the "..." tab
And I should see the tab "..."
And I switch back to the first tab
And I switch back to the previous tab
And I switch to the "..." tab
And I switch to the newly opened tab
```

**Comment by triskweline** [comment]
I never had a test that needed to switch between tabs.

**Comment by judithroth** [comment]
I used this in more than one project I think. Sometimes you have links that open in a new tab and you want to test that / your test has to follow the link. But still I am not sure if it is too specific to add it to spreewald.

**Comment by makmic** [comment]
I'd love to see a nice set of tab-switching steps in Spreewald. Note that some of its use cases could also be solved [with multiple sessions](https://makandracards.com/makandra/474480-how-to-make-a-cucumber-test-work-with-multiple-browser-sessions).

**Comment by triskweline** [comment]
If we have multiple apps that use steps I'm fine with adding them in.

I'm using the session steps a lot in the app I'm currently working on (it's a multiuser chat), but I'm not super happy with the definitions' regexp (`... [session: foo]`).

**Comment by codener** [comment]
I believe tabbing steps are a) project-independent and b) complicated to construct correctly, so adding them to Spreewald would be a real advantage. It enables any project to just use them instead of having to reinvent them.

**Comment by makmic** [comment]
@denzelem I just started the implementation of these steps and wanted to clarify some of the steps you mentioned.
* What is the difference between `I open the "[^"]+" tab` and `I switch to the "[^"]+" tab` ?
 I'd assume that we can drop the former.
* Are  `I switch to the "[^"]+" tab` and `I should see the tab "[^"]+"` referencing a tab's _title_?
* is `I open (.+?) in a new tab` referencing a page name (of "paths.rb") or a visible element of that name? (I'm assuming the former)
* `I switch back to the first tab`  seems to be an alias to `I switch back to the previous tab`. 
In this case I'd drop the first step - ok?

**Comment by denzelem** [comment]
> What is the difference between I open the "[^"]+" tab and I switch to the "[^"]+" tab ?
I'd assume that we can drop the former.

Actually the first one didn't belong to this issue. It is a step from a project that means the navigation tab, so it follows a link.

> Are I switch to the "[^"]+" tab and I should see the tab "[^"]+" referencing a tab's title?

Sorry, both do not belong to the browser tab, but the navigation tab.

> is I open (.+?) in a new tab referencing a page name (of "paths.rb") or a visible element of that name? (I'm assuming the former)

Yes you are right. 

Exampel: `And I open the homepage in a new tab`

```
When(/^I open (.*?) in a new tab$/) do |page_name|
  @first_window_handle = page.driver.current_window_handle
  path = path_to(page_name)
  id = "helper_#{SecureRandom.hex(8)}"
  execute_script <<-JAVASCRIPT
    var $helper = $('<a>').attr({ href: #{path.to_json}, id: #{id.to_json}, target: '_blank' });
    $helper.prependTo('body').text('click me').css({ zIndex: 9999, position: 'absolute' });
    setTimeout(function() { $helper.remove() }, 2000); // Autoremove helper link
  JAVASCRIPT
  find("##{id}").click
  step 'I switch to the newly opened tab'
end
```

> I switch back to the first tab seems to be an alias to I switch back to the previous tab.
In this case I'd drop the first step - ok?

Yes it seems to be like this. Even if `I switch back to the first tab` sound for me different.

```
Before do
  @first_window_handle = nil
end

When(/^I open (.*?) in a new tab$/) do |page_name|
  @first_window_handle = page.driver.current_window_handle
  path = path_to(page_name)
  id = "helper_#{SecureRandom.hex(8)}"
  execute_script <<-JAVASCRIPT
    var $helper = $('<a>').attr({ href: #{path.to_json}, id: #{id.to_json}, target: '_blank' });
    $helper.prependTo('body').text('click me').css({ zIndex: 9999, position: 'absolute' });
    setTimeout(function() { $helper.remove() }, 2000); // Autoremove helper link
  JAVASCRIPT
  find("##{id}").click
  step 'I switch to the newly opened tab'
end

When('I switch back to the first tab') do
  (Capybara.current_driver == :selenium) || raise('This step works only with selenium')
  page.driver.browser.switch_to.window(@first_window_handle)
end
```

```
When(/^I switch back to the previous tab$/) do
  (Capybara.current_driver == :selenium) || raise('This step works only with selenium')
  # Second last should be the previous tab (Last is the current one)
  page.driver.browser.switch_to.window(page.driver.browser.window_handles[-2])
end
```

**Comment by makmic** [comment]
Thanks for your valuable feedback! I think the list of actually implemented tab steps will look sth. like this:

```
When /^I open (.+?) in a new browser tab$/
When 'I close the browser tab'
When /^I switch to the new(?:ly opened) browser tab$/
When /^I switch(?: back)? to the previous browser tab$/
When 'I may open a new browser tab' # required for "I should/not have opened .."

Then /^I should(not )? have opened a new browser tab$/
Then /^I should see \d+ browser tabs$
Then /^I should see at least \d+ browser tabs$/
```

**Comment by makmic** [comment]
We just released Spreewald 2.5.0 which includes the steps mentioned above :partying_face:
